### DISCUSSION TITLE
Activating SumatraPDF from AutoHotKey?

### DISCUSSION THREAD

**Discussion by ruinatokyo** [sumatrapdfreader/sumatrapdf#4376] [Documentation]
Sumatra PDF has behavior different from every other app I use. I have an AHK script to toggle activation of windows (for example, one key to switch between a reference PDF and the Word document I am editing). The script calls the app for activation by its exe ("SumatraPDF.exe"), but Sumatra does not come to the front and seems to not come into focus (as shortcut keys and the like do not work until I physically click the Sumatra window).

I believe I also tried by pid and had the same problem. Is there a surefire way to activate and give focus to Sumatra PDF?

**Comment by GitHubRulesOK** [comment]
It should be accessible same as others but it may be running a slightly different way as it has to be able to broadcast its self controls via DDE

Whenever I have tried driving with AHK GUI menus in the past it worked flawlessly
E.G.  for swap keys or sending parts of books to the web or read aloud or select line or pick colour from pallet to highlight text

Most of those features are now to some degree in main app, so have not run AHK for months so it may have changed ?

For what it is worth one of my old test files is at https://github.com/sumatrapdfreader/sumatrapdf/issues/217#issuecomment-596605107
It is not trying to set focus only find the SumatraPDF session "already in focus."
It does not do much ! but if SumatraPDF is running it will still behave as designed so as to shift character selection in middle of a word to the start of word so you can drag the following text inclusive of that word ! 

Some SumatraPDF staring behaviours may have changed since then but if you are calling the exe via shell or DDE it should launch in a similar fashion.  I dont know if there may have been some changes in Windows Focus or SumatraPDF response to some newer Windows calls, but it should focus with calling the exe/dde server.

You are of course using AHK32 bit for SumatraPDF32.exe and 64 for 64  ? You can not expect to breach UAC with a 32/64 mix

SumatraPDF is daily brought into "Focus" (probably thousands of times) from LaTeX editors or pirate wrappers ! by calling the exe via command line or DDE the key to that command line call is -reuse-instance to recall with the named PDF but there are many other calls.

**Comment by kjk** [comment]
Let's start with: what exactly is your script doing? Post the script because `calls the app for activation by its exe` doesn't tell me what actually happens.

Generally speaking, AHK can call any windows API so I'm sure you can solve that using AHK by grabbing Sumatra window handle and calling the necessary windows API functions to bring it to front, give focus etc.

But I'm not AHK person nor do I know what exactly you're trying to achieve so can't tell you what the script would be.

**Reply by ruinatokyo** [reply]
Thanks for your reply. This is an excerpt of the method I use. I suppose I will in the end have to try the Windows API method.

```autohotkey
; Set target window to Sumatra PDF
>!SC03A:: ; hotkey to set target window
    MsgBox, Activate Sumatra PDF window.
    KeyWait, LButton, D
    Sleep, 300
    WinGet, tgt1, ProcessName, A
    tgt := tgt1
    winTitle2 := "ahk_exe " . tgt
Return

; Toggle to target window (Sumatra PDF)
SC03A:: ; hotkey to switch to target window
    ; If target is not set, exit
    if (tgt = "")
    {
        MsgBox, Please set the target window first.
        return
    }

    ; Activate Sumatra PDF
    WinActivate %winTitle2%
Return
```

**Comment by kjk** [comment]
I don't really know AHK but:
```
winTitle2 := "ahk_exe " . tgt
...
WinActivate %winTitle2%
```
I don't see how `winTitle2` could possibly be the name of Sumatra window hence I don't see how `WinActivate` would work.

**Reply by GitHubRulesOK** [reply]
@kjk .tgt should in theory be SumatraPDF.exe (session.exe) e.g. `Hotkey, "IfWinActive", "ahk_exe SumatraPDF.exe"`
but commonly may be called via class `Hotkey, IfWinActive, ahk_class SUMATRA_PDF_FRAME`

My sample uses the later and is based on SumatraPDF already running so seems to be unchanged since the last time I used it to still working in current pre-release. Win Activate sounds like VBS method to start via com

**Reply by ruinatokyo** [reply]
@kjk 
ahk_exe (Process Name/Path)
Use ahk_exe ProcessNameOrPath in WinTitle to identify a window belonging to any process with the given name or path.
This works with every app I have encountered, except Sumatra PDF.

@GitHubRulesOK
Thanks. I added an exception to the script to call Sumatra alone by ahk_class and it is working now. Hopefully the frame name doesn't change ;)

**Reply by GitHubRulesOK** [reply]
@ruinatokyo  I seem to recollect (too long ago without trawling a 3rd party similar issue)  my own reason to use class not exe was because SumatraPDF traditionally varies the title multiple ways so only one is convention and I "think" that's the reason its "different"

**Comment by Binocular222** [comment]
Don't use ahk_exe.  Please use ahk_class, i.e:
WinActivate ahk_class SUMATRA_PDF_FRAME

Refer to [online manual](https://www.autohotkey.com/docs/v1/misc/WinTitle.htm#ahk_class)
